inc/azure_macro_utils/macro_utils.h

```diff
@@ -202,6 +202,61 @@ const char* MU_C3(MU_, enumIdentifier,_ToString)(enumIdentifier value)
 
 #define MU_ENUM_VALUE_2(enumIdentifier, value) "", MU_ENUM_TO_STRING_2(enumIdentifier, (value)), (int)(value)
 
+#define CONVERT_ENUM_VALUE(from, to) \
+    case (from): \
+        *enum_value_to = (to); \
+        return 0; \
+
+// This macro declares an enum conversion function that translates a value from one enum to another
+#define MU_DECLARE_CONVERT_ENUM(ENUM_TYPE_FROM, ENUM_TYPE_TO) \
+    int MU_C4(convert_enum_, ENUM_TYPE_FROM, _, ENUM_TYPE_TO)(ENUM_TYPE_FROM enum_value_from, ENUM_TYPE_TO* enum_value_to);
+
+// This macro defines an enum conversion function that translates a value from one enum to another
+// It is implemented as a switch statement. This allows for catching multiple from values being in the list also
+#define MU_DEFINE_CONVERT_ENUM_WITHOUT_INVALID(ENUM_TYPE_FROM, ENUM_TYPE_TO, ...) \
+    int MU_C4(convert_enum_, ENUM_TYPE_FROM, _, ENUM_TYPE_TO)(ENUM_TYPE_FROM enum_value_from, ENUM_TYPE_TO* enum_value_to) \
+    { \
+        switch (enum_value_from) \
+        { \
+            default: \
+                break; \
+            MU_FOR_EACH_2(CONVERT_ENUM_VALUE, __VA_ARGS__) \
+        } \
+        return MU_FAILURE; \
+    }
+
+// This macro simply adds the conversion of the "well known" by now _INVALID enum value
+#define MU_DEFINE_CONVERT_ENUM(ENUM_TYPE_FROM, ENUM_TYPE_TO, ...) \
+    MU_DEFINE_CONVERT_ENUM_WITHOUT_INVALID(ENUM_TYPE_FROM, ENUM_TYPE_TO, MU_C2(ENUM_TYPE_FROM, _INVALID), MU_C2(ENUM_TYPE_TO, _INVALID), __VA_ARGS__)
+
+#define CONSTRUCT_FROM_FAKE_ENUM(from_value, to_value) \
+    MU_C3(fake_from_, from_value, _C0FB3709_EDE8_4288_8F70_FDDB5D8D7A51),
+
+#define CONSTRUCT_TO_FAKE_ENUM(from_value, to_value) \
+    MU_C3(fake_to_, to_value, _C0FB3709_EDE8_4288_8F70_FDDB5D8D7A51),
+
+// This macro adds validation of the fact that the from values count in the list match the count of values in the
+// from enum type. It does that by having a special fake enum where all the "from" values are enumerated
+// The count check is done by having a zero sized array in a struct if the counts do not match
+#define MU_DEFINE_CONVERT_ENUM_WITH_VALIDATION(ENUM_TYPE_FROM, ENUM_TYPE_TO, ...) \
+    /* Have an enum which has all the "from" values */ \
+    /* This enum will have a count */ \
+    typedef enum MU_C3(fake_, ENUM_TYPE_FROM, _C0FB3709_EDE8_4288_8F70_FDDB5D8D7A51_TAG) \
```

Comment:
>typedef enum [](start = 4, length = 12)

just wondering if this could have been constructed using MU_DEFINE_ENUM.

Reply:
Done with a MU_DEFINE_ENUM.

---
In reply to: [394386401](https://github.com/Azure/azure-macro-utils-c/pull/36#discussion_r394386401) [](ancestors = 394386401)

Reply:
Spoke too soon, I tried and I puffed and it turned annoying because expansion to one token, so I left it as is as it works.

---
In reply to: [394386401](https://github.com/Azure/azure-macro-utils-c/pull/36#discussion_r394386401) [](ancestors = 394386401)

Reply:
Pretty sure it can be done with enough sokoban, but I just did not feel into it now.

---
In reply to: [394635565](https://github.com/Azure/azure-macro-utils-c/pull/36#discussion_r394635565) [](ancestors = 394635565,394386401)